core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -36,12 +41,38 @@
  */
 public class CurrentIteration implements ValueExpression {
 
+    private final ValueExpression level;
+
+    public CurrentIteration(final ValueExpression level) {
+        this.level = checkNotNull(level, "level");
+    }
+
     @Override
     public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
-        if (parseState.iterations.head == null) {
-            return ImmutableList.create(Optional.empty());
+        return ImmutableList.create(getIteration(parseState, encoding));
+    }
+
+    private Optional<Value> getIteration(final ParseState parseState, final Encoding encoding) {
+        final BigInteger level = getLevel(parseState, encoding);
+        if (parseState.iterations.size <= level.longValue()) {
+            return Optional.empty();
+        }
+        return getIterationRecursive(parseState.iterations, level);
+    }
+
+    private BigInteger getLevel(final ParseState parseState, final Encoding encoding) {
+        final ImmutableList<Optional<Value>> evaluatedLevel = level.eval(parseState, encoding);
+        if (evaluatedLevel.size != 1 || !evaluatedLevel.head.isPresent()) {
+            throw new IllegalArgumentException("Level must evaluate to a single non-empty value.");
+        }
+        return evaluatedLevel.head.get().asNumeric();
+    }
+
+    private Optional<Value> getIterationRecursive(final ImmutableList<BigInteger> iterations, final BigInteger level) {
```

Comment:
This is a recursive method, so it should be implemented with a `Trampoline`. I agree it will probably never lead to scalability issues in practice, but let's avoid some weird corner case by always avoid hard recursion :)